OIIInspector/image_manager.py

```diff
@@ -0,0 +1,187 @@
+        """
+        port_start = self._grpc_start_port
+        port_end = port_start + self._grpc_max_port_tries
+
+        for port in range(port_start, port_end):
+            try:
+                return (
+                    port,
+                    self._serve_index_registry_at_port(
+                        port, self._grpc_init_wait_time),
+                )
+            except exceptions.AddressAlreadyInUse:
+                log.info('Port %d is in use, trying another.', port)
+
+        err_msg = f"No free port has been found after {self._grpc_max_port_tries} attempts."
+        raise exceptions.NoFreePortFound(err_msg)
+
+    @retry(exceptions=exceptions.OIIInspectorError, tries=2, logger=log)
+    def _serve_index_registry_at_port(self, port, wait_time):
+        """
+        Start an image registry service at a specified port.
+
+        :param str int port: port to start the service on.
+        :param wait_time: time to wait before checking if the service is initialized.
+        :return: object of the running Popen process.
+        :rtype: Popen
+        :raises OIIInspectorError: if the process has failed to initialize too many times, or an unexpected
+        error occurred.
+        :raises AddressAlreadyInUse: if the specified port is already being used by another service.
+        """
+        cmd = f"{self._container_platform} run --name={self._base_container_name}_{port} " \
+              f"-p={port}:{port} {self._image_address}"
+
+        rpc_proc = subprocess.Popen(
+            shlex.split(cmd),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+        )
+        start_time = time.time()
+        while time.time() - start_time < wait_time:
+            time.sleep(1)
+            ret = rpc_proc.poll()
+            # process has terminated
+            if ret is not None:
+                stderr = rpc_proc.stderr.read()
+                if "address already in use" in stderr:
+                    raise exceptions.AddressAlreadyInUse(
+                        f"Port {port} is already used by a different service")
+                if "the container name" in stderr and "is already in use by" in stderr:
+                    raise exceptions.AddressAlreadyInUse(
+                        f"Port {port} is already used by service by same image name: {self._base_container_name}")
+                raise exceptions.OIIInspectorError(f"Command {cmd} has failed with error {stderr}"
+                                                   .format(cmd="".join(cmd), stderr=stderr))
+
+            # query the service to see if it has started
+            try:
+                output = run_cmd(f"grpcurl -plaintext localhost:{port} list api.Registry")
+            except RuntimeError:
+                log.info(f"grpcurl -plaintext localhost:{port} list api.Registry failed")
```

Comment:
We probably shouldn't print this, right? It might be printed many times between attempts